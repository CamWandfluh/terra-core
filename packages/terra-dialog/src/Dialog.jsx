import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import 'terra-base/lib/baseStyles';
import Button from 'terra-button';
import ContentContainer from 'terra-content-container';
import styles from './Dialog.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The children to be placed within the main content area of the dialog.
   */
  children: PropTypes.node,
  /**
   * React node to be placed within the header area of the dialog.
   */
  header: PropTypes.node.isRequired,
  /**
   * React node to be placed within the footer area of the dialog.
   * Content exceeding the available space will be hidden.
   */
  footer: PropTypes.node.isRequired,
  /**
   * Whether or not additional footer styling needs to be applied. Used in conjunction with custom content.
   */
  isFooterStyled: PropTypes.bool,
  /**
   * Callback function for when the close button is clicked. The close button will not display if this is not set.
   * On small viewports a back button will be displayed instead of a close button.
   */
  onClose: PropTypes.func,
  /**
   * Ref callback for the scrollable area of the content container.
   */
  scrollRefCallback: PropTypes.func,
};

const defaultProps = {
  onClose: null,
  children: null,
  isFooterStyled: false,
};

const contextTypes = {
  /* eslint-disable consistent-return */
  intl: (context) => {
    if (context.intl === undefined) {
      return new Error('Please add locale prop to Base component to load translations');
    }
  },
};

const Dialog = ({ children, header, footer, isFooterStyled, onClose, scrollRefCallback, ...customProps }, { intl }) => {
  const dialogClassNames = cx([
    'dialog',
    customProps.className,
  ]);

  const closeText = intl.formatMessage({ id: 'Terra.dialog.close' });
  const closeButton = onClose ? <div className={cx('dialog-header-close')}><Button variant="utility" text={closeText} onClick={onClose} isIconOnly icon={<span className={cx('close-icon')} />} /></div> : null;

  const dialogHeader = <div className={cx('dialog-header')}><div className={cx('dialog-header-title')}>{header}</div>{closeButton}</div>;
  const dialogFooter = isFooterStyled ? <div className={cx('dialog-footer')}>{footer}</div> : footer;

  return (
    <ContentContainer
      {...customProps}
      className={dialogClassNames}
      header={dialogHeader}
      footer={dialogFooter}
      scrollRefCallback={scrollRefCallback}
      fill
    >
      {children}
    </ContentContainer>
  );
};

Dialog.propTypes = propTypes;
Dialog.defaultProps = defaultProps;
Dialog.contextTypes = contextTypes;

export default Dialog;
