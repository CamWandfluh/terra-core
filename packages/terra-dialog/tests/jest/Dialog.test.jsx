import React from 'react';
import intlContexts from './intl-context-setup';
import Dialog from '../../src/Dialog';

describe('Dialog', () => {
  const defaultRender = (
    <Dialog
      header="Header Content"
      footer="Footer Content"
      scrollRefCallback={jest.fn()}
    >
      some body content
    </Dialog>
  );

  // Snapshot Test
  it('should render a default component', () => {
    const wrapper = shallow(defaultRender, intlContexts.shallowContext);
    expect(wrapper).toMatchSnapshot();
  });

  // Structure Tests
  it('should have the class dialog', () => {
    const wrapper = shallow(defaultRender, intlContexts.shallowContext);
    expect(wrapper.prop('className')).toContain('dialog');
  });

  it('should render a Dialog with merged attributes', () => {
    const wrapper = shallow(<Dialog header="Header Content" footer="Footer Content" className="TestClass">some body content</Dialog>, intlContexts.shallowContext);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a Dialog with props - isStyledFooter', () => {
    const wrapper = shallow(<Dialog header="Header Content" footer="Footer Content" isStyledFooter>some body content</Dialog>, intlContexts.shallowContext);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a Dialog with props - onClose', () => {
    const wrapper = shallow(<Dialog header="Header Content" footer="Footer Content" onClose={jest.fn()}>some body content</Dialog>, intlContexts.shallowContext);
    expect(wrapper).toMatchSnapshot();
  });
});

it('throws error on missing locale prop in Base', () => {
  try {
    render(<Dialog header="Header Content" footer="Footer Content" />);
  } catch (e) {
    expect(e.message).toContain('add locale prop to Base component');
  }
});
