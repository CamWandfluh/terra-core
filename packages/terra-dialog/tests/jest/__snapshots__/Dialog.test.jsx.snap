// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog should render a Dialog with merged attributes 1`] = `
<ContentContainer
  className="dialog TestClass"
  fill={true}
  footer="Footer Content"
  header={
    <div
      className="dialog-header"
    >
      <div
        className="dialog-header-title"
      >
        Header Content
      </div>
    </div>
  }
>
  some body content
</ContentContainer>
`;

exports[`Dialog should render a Dialog with props - isStyledFooter 1`] = `
<ContentContainer
  className="dialog"
  fill={true}
  footer="Footer Content"
  header={
    <div
      className="dialog-header"
    >
      <div
        className="dialog-header-title"
      >
        Header Content
      </div>
    </div>
  }
  isStyledFooter={true}
>
  some body content
</ContentContainer>
`;

exports[`Dialog should render a Dialog with props - onClose 1`] = `
<ContentContainer
  className="dialog"
  fill={true}
  footer="Footer Content"
  header={
    <div
      className="dialog-header"
    >
      <div
        className="dialog-header-title"
      >
        Header Content
      </div>
      <div
        className="dialog-header-close"
      >
        <Button
          icon={
            <span
              className="close-icon"
            />
          }
          isBlock={false}
          isCompact={false}
          isDisabled={false}
          isIconOnly={true}
          isReversed={false}
          onClick={[MockFunction]}
          text="Close"
          type="button"
          variant="utility"
        />
      </div>
    </div>
  }
>
  some body content
</ContentContainer>
`;

exports[`Dialog should render a default component 1`] = `
<ContentContainer
  className="dialog"
  fill={true}
  footer="Footer Content"
  header={
    <div
      className="dialog-header"
    >
      <div
        className="dialog-header-title"
      >
        Header Content
      </div>
    </div>
  }
  scrollRefCallback={[MockFunction]}
>
  some body content
</ContentContainer>
`;
